File: src/java/io/BufferedInputStream.cpp
#include "java/io/BufferedInputStream.h"
#include "java/io/FileInputStream.h"

namespace java {
namespace io {

BufferedInputStream::BufferedInputStream(InputStream *inputStream, bool ownInputStream):
    inputStream(inputStream),
    ownInputStream(ownInputStream)
{
}

BufferedInputStream::~BufferedInputStream() {
    dispose();
}

bool
BufferedInputStream::isOpen() const {
    if ( inputStream == nullptr ) {
        return false;
    }
    const FileInputStream *fileInputStream = dynamic_cast<const FileInputStream *>(inputStream);
    return fileInputStream != nullptr ? fileInputStream->isOpen() : true;
}

int
BufferedInputStream::read() {
    if ( inputStream == nullptr ) {
        return -1;
    }
    return inputStream->read();
}

int
BufferedInputStream::read(unsigned char *buffer, int offset, int length) {
    if ( inputStream == nullptr ) {
        return -1;
    }
    return inputStream->read(buffer, offset, length);
}

long
BufferedInputStream::tell() const {
    if ( inputStream == nullptr ) {
        return -1;
    }
    const FileInputStream *fileInputStream = dynamic_cast<const FileInputStream *>(inputStream);
    return fileInputStream != nullptr ? fileInputStream->tell() : -1;
}

void
BufferedInputStream::close() {
    dispose();
}

void
BufferedInputStream::dispose() {
    if ( inputStream == nullptr ) {
        return;
    }
    inputStream->dispose();
    if ( ownInputStream ) {
        delete inputStream;
    }
    inputStream = nullptr;
}

}
}

File: typescript/src/main/java/io/BufferedInputStream.ts
import { InputStream } from "./InputStream";

export class BufferedInputStream extends InputStream {
  protected inputStream: InputStream | null;

  public constructor(inputStream: InputStream | null = null) {
    super();
    this.inputStream = inputStream;
  }

  public read(): number;
  public read(buffer: Uint8Array, offset: number, length: number): number;
  public read(buffer?: Uint8Array, offset?: number, length?: number): number {
    if (this.inputStream === null) {
      return -1;
    }

    if (buffer === undefined) {
      return this.inputStream.read();
    }
    return this.inputStream.read(buffer, offset ?? 0, length ?? 0);
  }

  public close(): void {
    this.dispose();
  }

  public override dispose(): void {
    if (this.inputStream === null) {
      return;
    }
    this.inputStream.dispose();
    this.inputStream = null;
  }
}
